### PR TITLE
An Overview of Docker Container Linking

### LINKED ISSUE #2786: An Overview of Docker Container Linking
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
An Overview of Docker Container Linking

### Introduction paragraph (2-3 paragraphs):
Docker consists of a linking system that allows multiple containers to be linked together. This linking system allows connection information to be sent from a source container to a recipient container.   

This article will take readers through Docker container linking. It will provide an overview of how Docker container linking works and explain how communication works across links. 

### Key takeaways: 
1. Introduction to Docker container linking
2. How Docker container linking works
3. Communication across links (How it works)
-environmental variables
-host file
4. Conclusion

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by artmuth**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2786

### CODE REVIEW

**Comment by mbaabuones** [comment]
@artmuth It seems you uploaded your files using the old structure. In your forked repository, click on fetch upstream->'fetch and merge' to update to the new structure. Delete your uploaded files and upload them in the 'articles' folder inside the 'content' folder. You can create a new PR.

**Comment by artmuth** [comment]
> @artmuth It seems you uploaded your files using the old structure. In your forked repository, click on fetch upstream->'fetch and merge' to update to the new structure. Delete your uploaded files and upload them in the 'articles' folder inside the 'content' folder. You don't have to create a new PR.

@mbaabuones Thank you. I have done that.

File: content/articles/an-overview-of-docker-container-linking/index.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+All the environmental variables that originate from the source container are exposed to the target container. This enables the target container to discover information relating to the source container. 
+
+##### Updating the /etc/hosts file
+Apart from sharing the environmental variables, Docker provides a host entry to the /etc/hosts file in regards to the source container. Two entries are added in the /etc/hosts file. 
+
+The first one is for the recipient container, which consists of a hostname in the form of container ID. The second entry employs the link identity to provide information of the source container (IP address). 
+
+This information is updated automatically if the source container is restarted. The recipient container will be provided with a reference to the new information of the source container, which enhances the continuity of linked communication.
```
Comment (mbaabuones) [inline content/articles/an-overview-of-docker-container-linking/index.md:48]:
```suggestion
This information is updated automatically if the source container is restarted. If this is the case, the recipient container will be provided with a reference to the new information of the source container. This enhances the continuity of linked communication.

File: content/articles/an-overview-of-docker-container-linking/index.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+##### Updating the /etc/hosts file
+Apart from sharing the environmental variables, Docker provides a host entry to the /etc/hosts file in regards to the source container. Two entries are added in the /etc/hosts file. 
+
+The first one is for the recipient container, which consists of a hostname in the form of container ID. The second entry employs the link identity to provide information of the source container (IP address). 
+
+This information is updated automatically if the source container is restarted. The recipient container will be provided with a reference to the new information of the source container, which enhances the continuity of linked communication. 
+
+### Conclusion
+This article has provided the basics of Docker container linking. The following are the main points we have learned:
```
Comment (mbaabuones) [inline content/articles/an-overview-of-docker-container-linking/index.md:51]:
```suggestion
This article has provided the basics of Docker container linking. The following are the main points we have learned:
1. Docker container linking allows multiple containers to be linked to each other. 
2. It allows the recipient container to get connection information relating to the source container.
3. You should use Docker container linking when you are using default bridge networks and when you want to share environmental variables.
4. Docker container linking works through two main processes: the naming of containers and communication across links.
5. Communication across links is achieved through two main ways: sharing environmental variables and updating the /etc/hosts file.

Happy learning!

File: content/articles/an-overview-of-docker-container-linking/index.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+Docker bridge networks employ a software bridge that enhance communication between containers that are connected to it. The software bridge prevents containers that are connected to it from getting or establishing communication. This means that containers using different bridges can communicate with each other. 
+
+Docker container linking should not be used when you are using user-defined bridges. These bridges are utilized in Docker networking. 
+
+#### When you want to share environmental variables
+Docker default bridge networks are the ideal networks for sharing environmental variables from one container to another. This is because the user-defined networks do not support the sharing of these variables. 
+
+### How Docker container linking works
+Docker container linking works through two main process: naming of containers and communication across links.
```
Comment (mbaabuones) [inline content/articles/an-overview-of-docker-container-linking/index.md:28]:
```suggestion
Docker container linking works through two main processes: the naming of containers and communication across links.

File: content/articles/an-overview-of-docker-container-linking/index.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+##### Environmental variables
+Docker can enhance communication between containers through the sharing of environmental variables from the source container to the target container. When a link is created, Docker generates many environmental variables (automatically) in the recipient container depending on the --link parameters that were passed. 
+
+All the environmental variables that originate from the source container are exposed to the target container. This enables the target container to discover information relating to the source container. 
+
+##### Updating the /etc/hosts file
+Apart from sharing the environmental variables, Docker provides a host entry to the /etc/hosts file in regards to the source container. Two entries are added in the /etc/hosts file. 
+
+The first one is for the recipient container, which consists of a hostname in the form of container ID. The second entry employs the link identity to provide information of the source container (IP address).
```
Comment (mbaabuones) [inline content/articles/an-overview-of-docker-container-linking/index.md:46]: a container ID

**Review by mbaabuones — COMMENTED** [review]
@artmuth Great article! Your article is well structured and informative. Please ensure you check for grammar errors before you submit your article. Work on the minor changes I have suggested to you and ping me when you are done.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
@artmuth 
Upon a quick glance - this article seems a bit light on substance - lets be sure to go above and beyond any official docs and/or other similar blog posts (to avoid possible article closure). 

cc @mbaabuones

**Comment by artmuth** [comment]
> @artmuth
> Upon a quick glance - this article seems a bit light on substance - lets be sure to go above and beyond any official docs and/or other similar blog posts (to avoid possible article closure).
> 
> cc @mbaabuones

@hectorkambow Thank you for the heads up. I have revised the article to add value. 
cc @mbaabuones

**Comment by artmuth** [comment]
@mbaabuones I have made the suggested changes. Please check.

**Review by mbaabuones — APPROVED** [review]

**Review by Kayere — APPROVED** [review]
Great article @artmuth, very easy to follow through.

Thank you for the review @mbaabuones.

**Pull request merged**